### PR TITLE
Add a primitive to compute the "transitive" heap size of a value

### PR DESCRIPTION
**Pull request by alainfrisch**
It is sometimes useful to know how "big" values are.  This can serve as a debugging tool (to track the cause of excessive memory usage), but also in production to adjust cache heuristics (e.g. eject "big-but-cheap-to-compute" values sooner).  It can also be useful to help people get a good feeling of how to design compact data structures and how not to break sharing of inner nodes.

So I propose to add such a function to the Obj module (to make it a bit scary to use).  It computes the total size (in words) of blocks accessible from a given value.  Statically allocated blocks are excluded (this can also serve to implement non-regression tests that check that a value is statically allocated).

The implementation was inspired by the technique used in extern.c: when a block is first traversed, its color is turned to "blue".  We maintain a queue of pointers which serves as a worklist (blocks yet to inspect) and also to restore original colors.  One word of internal storage is required per visited block.  (I had a first implementation reusing data structures from extern.c, but they use at least two words per block.)

If this proposal is well received, I'll write some unit tests. [EDIT: done]

### CODE REVIEW

**Comment by dbuenzli** [comment]
[MPR #6820](http://caml.inria.fr/mantis/view.php?id=6829)

**Comment by alainfrisch** [comment]
Any comment on the proposal?  Is `Obj` the appropriate module to put this function?  I'd also appreciate if another pair of eyes could review the code; it's easy to make a mistake...

**Comment by dbuenzli** [comment]
> Is Obj the appropriate module to put this function?

I think it's better than the `Gc` module proposed in MPR #6820.

**Comment by mshinwell** [comment]
I only had a very quick look (and am jetlagged) but it looks like caml_stat_alloc should probably be used instead of malloc, no?

I'm a bit torn about Gc versus Obj.  In some sense, if this is going to be a safe operation, there's no need for it to be in Obj.  However it doesn't seem to quite fit into Gc either.

**Comment by alainfrisch** [comment]
> I only had a very quick look (and am jetlagged) but it looks like caml_stat_alloc should probably be used instead of malloc, no?

I got inspired by extern.c which uses malloc.  A quick grep-fight in byterun gives an inconclusive:

```
$ grep malloc *.c  | wc -l
34

$ grep caml_stat_alloc *.c  | wc -l
32
```

Does anyone know how to choose?

>  In some sense, if this is going to be a safe operation,

`Obj.size` is also safe (I think).  But such function are very implementation dependent.  Typically, it's not clear than javascript backend will support reachable_words (with a close-enough semantics -- I'm thinking about their ability to look into closures).  Perhaps `Sys`?

File: byterun/obj.c [hunk 1]
```diff
@@ -260,3 +260,115 @@ CAMLprim value caml_fresh_oo_id (value v) {
...
+     An interesting block is block in the heap which does not
+     represent an infix pointer. Infix pointers are normalized to the
+     beginning of their block.  Blocks in the static data area are excluded.
+
+     When a block is inserted in the queue, it is marked as blue.
+     This mark is used to avoid a second visit of the same block.
+     The real color is stored in the last 2 bits of the pointer in the
+     queue.  (Same technique as in extern.c.)
+
+     The queue is stored as a linked list of blocks.  Initially,
```
Comment (damiendoligez) [inline byterun/obj.c]: This line is confusing because there are two kinds of blocks.

File: byterun/obj.c [hunk 2]
```diff
@@ -260,3 +260,115 @@ CAMLprim value caml_fresh_oo_id (value v) {
...
+  while (read_pos != write_pos || read_block != write_block) {
+    /* Pop the next element from the queue */
+    if (read_pos == ENTRIES_PER_QUEUE_BLOCK) {
+      read_pos = 0;
+      read_block = read_block->next;
+    }
+    v = read_block->entries[read_pos++] & ~3;
+
+    hd = Hd_val(v);
+    sz = Wosize_hd(hd);
```
Comment (damiendoligez) [inline byterun/obj.c:326]: You should use `Whsize_hd` instead of adding 1 by hand.

**Comment by damiendoligez** [comment]
I think `Obj` is better than `Sys`.

File: byterun/obj.c (hunk 2 above)
Comment (alainfrisch) [inline byterun/obj.c:326]: I'll use `whsize_wosize(sz)` since `sz` is already bound.  Is that ok?

**Comment by alainfrisch** [comment]
@damiendoligez Can you comment about malloc vs caml_stat_alloc?

**Comment by lefessan** [comment]
I am really surprised by this merge. I was thinking that we were waiting for Damien's feedback, and that the technical discussion was not over, as I think the primitive you described is far from being as useful as you said, in its current state (at least, for a cache or whatever, I would expect the primitive to come with a threshold argument, that would stop the traversal if the size becomes greater than the threshold).
@alainfrisch @damiendoligez

**Comment by alainfrisch** [comment]
Damien posted some (minor) comments, which I addressed.  I guess that he would have said so if he was against the proposal.

The threshold is indeed a good idea for use in caches, but the current version is already very useful for debugging purposes.

**Comment by DemiMarie** [comment]
@alainfrisch I think that one big difference is that if the runtime becomes unloadable, then `caml_stat_alloc` will use a memory pool to ensure all allocations are freed.

**Comment by DemiMarie** [comment]
CI fails with flambda.

**Comment by damiendoligez** [comment]
@drbo This means we should be using `caml_stat_alloc` everywhere instead of `malloc`, right?

@alainfrisch minor comments don't really mean that I approve of everything else, especially when I don't add the "approved" label...

**Comment by DemiMarie** [comment]
@damiendoligez I would imagine so (not part of the core team).

**Comment by yakobowski** [comment]
FWIW, we (the Frama-C team) have been using @backtracking's [size](https://www.lri.fr/~filliatr/ftp/ocaml/ds/size.mli.html) library for a long time. This is more or less what is proposed here. Our experience is that more functions are needed very rapidly. In particular, we often need to compute  "the size of _this_ value except what is also reachable from _that_ value".

**Comment by alainfrisch** [comment]
> In particular, we often need to compute "the size of this value except what is also reachable from that value".

This would be quite easy to implement, but costly if "that" value itself is big.  A variant is to "cut" only one (or several) specific values (but not blocks accessible from them).

**Comment by xavierleroy** [comment]
> This means we should be using caml_stat_alloc everywhere instead of malloc, right?

Not always.  Only if you're happy with caml_stat_alloc aborting the current function if it runs out of memory.  However, this can lead to memory leaks.  Consider two out-of-heap allocations in sequence:

```
  char * buf1 = caml_stat_alloc(size1);
  char * buf2 = caml_stat_alloc(size2);
```

If the second allocation runs out of memory (but not the first), an exception is raised and the space for  `buf1` will never be released.  In such cases, it is better to write

```
  char * buf1 = malloc(size1);
  if (buf1 == NULL) caml_raise_out_of_memory();
  char * buf2 = malloc(size2);
  if (buf2 == NULL) { free(buf1); caml_raise_out_of_memory(); }
```

**Comment by damiendoligez** [comment]
OK, so don't use `caml_stat_alloc` as a replacement for `malloc`. We'll have to review all uses of `malloc` to implement the unloadable runtime anyway.

**Pull request merged**